clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -678,30 +686,73 @@ constructKernelName(Sema &S, FunctionDecl *KernelCallerFunc,
 namespace {
 
 QualType getItemType(const FieldDecl *FD) { return FD->getType(); }
-QualType getItemType(const CXXBaseSpecifier &BS) { return BS.getType(); }
+// FIXME enable this function when base classes are supported
+// QualType getItemType(const CXXBaseSpecifier &BS) { return BS.getType(); }
 
 // Implements the 'for-each-visitor'  pattern.
 template <typename ParentTy, typename... Handlers>
 static void VisitAccessorWrapper(CXXRecordDecl *Owner, ParentTy &Parent,
                                  CXXRecordDecl *Wrapper,
                                  Handlers &... handlers);
 
+template <typename RangeTy, typename... Handlers>
+static void VisitField(CXXRecordDecl *Owner, RangeTy Item, QualType ItemTy,
+                       Handlers &... handlers) {
+  if (Util::isSyclAccessorType(ItemTy)) {
+    (void)std::initializer_list<int>{
+        (handlers.handleSyclAccessorType(Item, ItemTy), 0)...};
+  } else if (Util::isSyclStreamType(ItemTy))
+    (void)std::initializer_list<int>{
+        (handlers.handleSyclStreamType(Item, ItemTy), 0)...};
+  else {
+    if (ItemTy->isArrayType()) {
+      VisitArrayElements(Item, ItemTy, handlers...);
+    } else if (ItemTy->isStructureOrClassType()) {
+      VisitAccessorWrapper(Owner, Item, ItemTy->getAsCXXRecordDecl(),
+                           handlers...);
+    }
+  }
+}
+
+template <typename RangeTy, typename... Handlers>
+static void VisitStructField(CXXRecordDecl *Owner, RangeTy Item,
+                             QualType ItemTy, Handlers &... handlers) {
+  (void)std::initializer_list<int>{(handlers.enterField(Owner, Item), 0)...};
+  VisitField(Owner, Item, ItemTy, handlers...);
+  (void)std::initializer_list<int>{(handlers.leaveField(Owner, Item), 0)...};
+}
+
+template <typename RangeTy, typename... Handlers>
+static void VisitArrayElements(RangeTy Item, QualType FieldTy,
+                               Handlers &... handlers) {
+  if (!Util::containsStructType(FieldTy)) {
```

Comment:
This function isn't arranged very well.  The PlainSTructType is odd, and much of the work being done here is unnecessary.

Also, why are you splitting up the check of accessor-wrapper or field?  These should be happening together.

Finally, we have a fixed count, just do a for-loop.

Reply:
An idea of a PlainStructType is to distinguish between structs that contain SYCL special types such as accessors and samplers, and those that don't. For the former, we need to visit the SYCL special types and deal with them explicitly, while other structs just need a bulk transfer.

None of the work in this function is unnecessary. Deleting any line will cause array support to not work or be less general, and I thought we want to make array support quite general, such as allowing:
    struct  {
      double d;
      Accessor a_array[2];
    } array_s_a[2];

Reply:
It reproduces a large amount of work from previous functions, is my point.  The work here is unnecessary because it happens elsewhere.